pkg/api/v2/reporter/reporter.go

```diff
@@ -192,38 +193,180 @@ func (e ResourceReports) FilterByKind(kind string) ResourceReports {
+// This is used to iterate over the resources in a consistent order.
+// The reports are keyed by references to the resources, so are not sortable.
+// There is no unique key for a resource, so we use the string representation of the name/namespace as the key,
+// and collect all the resources with the same key together.
+// The previous implementation did not guarantee a consistent order when iterating over the resources,
+// so any order used here will be acceptable for backwards compatibility.
+func (e ResourceReports) refMapAndSortedKeys() (map[string][]resources.InputResource, []string) {
+	// refKeys contains all the unique keys for the resources
+	var refKeys []string
+	// refMaps is a map of resource keys to a slice of resources with that key
+	var refMap = make(map[string][]resources.InputResource)
+
+	// Loop over the resources
+	for res := range e {
+
+		// Get a string representation of the resource ref. This is not guaranteed to be unique.
+		resKey := res.GetMetadata().Ref().String()
+
+		// Add the key to the list of keys if it is not already there
+		if !slices.Contains(refKeys, resKey) {
+			refKeys = append(refKeys, resKey)
+		}
+		// Add the resource to the map of resources with the same key
+		refMap[resKey] = append(refMap[resKey], res)
+	}
+
+	// Sort the name-namespace keys. This will allow the reports to be accssed in a consistent order,
+	// except in cases where the name/namespace is not unique. In those cases, the individual validaiton
+	// functions will have to handle consistent ordering.
+	slices.Sort(refKeys)
+	return refMap, refKeys
+}
+
+// sortErrors sorts errors based on string representation
+// Note: because we are using multierror the string representation starts with "x errors occurred".
+// This will be consistent, but possibly counter-intuitive for tests.
+func sortErrors(errs []error) {
+	sort.Slice(errs, func(i, j int) bool {
+		return errs[i].Error() < errs[j].Error()
+	})
+}
+
+// Validate ignores warnings
 func (e ResourceReports) Validate() error {
 	var errs error
-	for res, rpt := range e {
-		if rpt.Errors != nil {
+	refMap, refKeys := e.refMapAndSortedKeys()
+
+	// the refKeys are sorted/sortable and point to the unsortable resources refs that are the keys to the report map
+	for _, refKey := range refKeys {
+		// name/namespace is not unique, so we collect those references together
+		reses := refMap[refKey]
+
+		var errsForKey []error
+		for _, res := range reses {
+			rpt := e[res]
+
+			if rpt.Errors != nil {
+				errsForKey = append(errsForKey, rpt.Errors)
+			}
+		}
+
+		if len(errsForKey) > 0 {
 			if errs == nil {
-				errs = errors.Errorf("invalid resource %v.%v", res.GetMetadata().Namespace, res.GetMetadata().Name)
+				// All of the resources in the group have the same metadata, so use the first one
+				errs = errors.Errorf("invalid resource %v.%v", reses[0].GetMetadata().Namespace, reses[0].GetMetadata().Name)
```

Comment:
nit: I understand that this is the way things have been, but this can be confusing as if there are multiple resources and errors in some of them, it will mention only the first resource and append errors belonging to other resources